uplc-syntax.md

```diff
@@ -34,12 +34,30 @@ module UPLC-CONCRETE-SYNTAX
                         | "bytestring"
                         | "unit"
                         | "bool"
+                        | "list" "(" TypeConstant ")"
+                        | "pair" "(" TypeConstant ")" "(" TypeConstant ")"
 
   syntax Constant ::= Int
                     | "True"
                     | "False"
                     | ByteString
                     | "()"
+                    | "[" ConstantList "]"
+                    | "(" Constant "," Constant ")"
+                    | "{" TextualData "}"
+
+  syntax ConstantList ::= List{Constant, ","}
+
+  syntax TextualData ::= "Constr" Int "[" DataList "]"
+                       | "Map" "[" DataPairList "]"
+                       | "List" "[" DataList "]"
```

Comment:
What inspired the choice for `[ ... ]` on these? Does that match some existing tool?

In Python maps are done with `{ ... }`, so maybe that would be one inspiration.

Also, should we consider just overloading the same operators for constants?

```
syntax Constant ::= "[" ConstantList "]" [klabel(valList), symbol, overload]

syntax TextualData ::= "[" DataList "]" [klabel(valList), symbol, overload]
```

Similar to what Wasm does here, this allows for a nice subsort heirarchy to be built.

Do we want a `ConstantList` to be a subsort of a `DataList`, for example? Or does that relationship make no sense? Probably the Plutus spec would tell us that.

Reply:
> What inspired the choice for `[ ... ]` on these? Does that match some existing tool?
> 
> In Python maps are done with `{ ... }`, so maybe that would be one inspiration.

The `uplc` unparser syntax is influenced by haskell. I believe this is why the list and pair values are enclosed in `[]` and `()` respectively.

For example, here's syntax that represents list of bools:

`(con list(bool) [True, False, True])`

and here's a program that represents a pair of integer and bool:

`(con pair (integer) (bool) (12345, True))`

So the values of the list are enclosed in `[]` and pairs are enclosed in `()`.

The syntax for the above 2 programs is supported by the `uplc` unparser.

### Reasons for choosing `[...]`

In thinking about a textual format for the `data` type, I decided to model this after how commonly used haskell tools will print out instances of the data ADT in haskell's interactive interpreter called `ghci`. in `ghci` you can load ADT definitions and print instances of these data types in an interactive prompt. The `Map` in the ADT isn't a builtin type called "map". Instead, the map is represented as a list of pairs (I'm assuming that the first element of the pair is a key and the second element is a value). Semantically, they might be equivalent but if we try to convert from this `Map` instance of `data` to a representation in untyped plutus using the `unMapData` builtin, we will get a term of type `list(pair(data) (data))`.

I hope this makes sense.. I'll respond to your latter comment in a separate response here.

Reply:
> Also, should we consider just overloading the same operators for constants?
> 
> ```
> syntax Constant ::= "[" ConstantList "]" [klabel(valList), symbol, overload]
> 
> syntax TextualData ::= "[" DataList "]" [klabel(valList), symbol, overload]
> ```
> 
> Similar to what Wasm does here, this allows for a nice subsort heirarchy to be built.
> 
> Do we want a `ConstantList` to be a subsort of a `DataList`, for example? Or does that relationship make no sense? Probably the Plutus spec would tell us that.

This is not valid because ConstantList can have types such as `bool` which is not supported by DataList.